### ISSUE TITLE
Unable to remove DRM from B&N epub, was working 10/8/2021

### ISSUE THREAD

**Issue by deliana-d** [apprenticeharper/DeDRM_tools#1929]
Hello! I just tried to convert some of my latest B&N purchases today, and it seems I am no longer able to remove the DRM via Claibre, though the same setup was working successfully about a month ago.

calibre Debug log
calibre 5.32 [64bit]  embedded-python: True is64bit: True
Windows-10-10.0.19041 Windows ('64bit', 'WindowsPE')
('Windows', '10', '10.0.19041')
Python 3.8.5
Windows: ('10', '10.0.19041', '', 'Multiprocessor Free')
Interface language: None
Successfully initialized third party plugins: DeDRM (7, 1, 0)
calibre 5.32 [64bit]  embedded-python: True is64bit: True
Windows-10-10.0.19041 Windows ('64bit', 'WindowsPE')
('Windows', '10', '10.0.19041')
Python 3.8.5
Windows: ('10', '10.0.19041', '', 'Multiprocessor Free')
Interface language: None
Successfully initialized third party plugins: DeDRM (7, 1, 0)
Turning on automatic hidpi scaling
devicePixelRatio: 1.0
logicalDpi: 96.0 x 96.0
physicalDpi: 135.5328125 x 135.46666666666664
Using calibre Qt style: True
[0.00] Starting up...
[0.00] Showing splash screen...
[1.52] splash screen shown
[1.52] Initializing db...
[1.53] db initialized
[1.53] Constructing main UI...
[2.42] main UI initialized...
[2.42] Hiding splash screen
Starting QuickView
DeDRM v7.1.0: Trying to decrypt 9781631230769.epub
DeDRM v7.1.0: Verifying zip archive integrity
DeDRM v7.1.0: “9781631230769.epub” is a secure Barnes & Noble ePub
DeDRM v7.1.0: Trying Encryption key mykey
Could not decrypt qjnkhznn.epub because of an exception:
Traceback (most recent call last):
  File "calibre_plugins.dedrm.ignobleepub", line 266, in decryptBook
  File "calibre_plugins.dedrm.ignobleepub", line 197, in __init__
  File "calibre_plugins.dedrm.ignobleepub", line 168, in __init__
  File "Crypto\Cipher\AES.py", line 232, in new
  File "Crypto\Cipher\__init__.py", line 79, in _create_cipher
  File "Crypto\Cipher\_mode_cbc.py", line 274, in _create_cbc_cipher
  File "Crypto\Cipher\AES.py", line 93, in _create_base_cipher
ValueError: Incorrect AES key length (0 bytes)

DeDRM v7.1.0: Failed to decrypt with key mykey after 0.1 seconds
DeDRM v7.1.0: Looking for new NOOK Study Keys after 0.1 seconds
No nook Study log files have been found.
DeDRM v7.1.0: Ultimately failed to decrypt after 0.1 seconds. Read the FAQs at Harper's repository: https://github.com/apprenticeharper/DeDRM_tools/blob/master/FAQs.md
Running file type plugin DeDRM failed with traceback:
Traceback (most recent call last):
  File "calibre\customize\ui.py", line 175, in _run_filetype_plugins
  File "calibre_plugins.dedrm.__init__", line 642, in run
  File "calibre_plugins.dedrm.__init__", line 311, in ePubDecrypt
calibre_plugins.dedrm.DeDRMError: DeDRM v7.1.0: Ultimately failed to decrypt after 0.1 seconds. Read the FAQs at Harper's repository: https://github.com/apprenticeharper/DeDRM_tools/blob/master/FAQs.md
[7.50] splash screen hidden
[7.50] Started up in 7.50 seconds with 118 books
stdout+stderr from file dialog helper: [b'', b'']
piped data from file dialog helper: [b'\x82\x06\xa3\xa9O\x1b\xca\x9a\xde\xe1\xfc\x9aT\xb1\x8c\xdem\xb9m\xd1\xcc\xcb\xd161\x18YV\xca\x80B\x9b', b'C:\\Users\\deliana\\AppData\\Local\\Packages\\BarnesNoble.Nook_ahnzqzva31enc\\LocalState\\9781631230769.epub']
Added Amaranthine Saga 05, FUMIKO v2 to db in: 0.1
Added 1 books in 0.9 seconds

Not sure if anything changed on the B&N side to cause this or what....

**Comment by ElleKayEm** [comment]
I've seen a couple reports lately where the B&N key needed to be extracted again because it had changed.

**Comment by ElleKayEm** [comment]
I don't think it necessarily matters in this case, but you're using an older version of DeDRM.  Current version is 7.2.1.

**Comment by deliana-d** [comment]
Yeah, tried the 7.2.1 just now as well and same result:

calibre Debug log
calibre 5.32 [64bit]  embedded-python: True is64bit: True
Windows-10-10.0.19041 Windows ('64bit', 'WindowsPE')
('Windows', '10', '10.0.19041')
Python 3.8.5
Windows: ('10', '10.0.19041', '', 'Multiprocessor Free')
Interface language: None
Successfully initialized third party plugins: DeDRM (7, 2, 1)
calibre 5.32 [64bit]  embedded-python: True is64bit: True
Windows-10-10.0.19041 Windows ('64bit', 'WindowsPE')
('Windows', '10', '10.0.19041')
Python 3.8.5
Windows: ('10', '10.0.19041', '', 'Multiprocessor Free')
Interface language: None
Successfully initialized third party plugins: DeDRM (7, 2, 1)
Turning on automatic hidpi scaling
devicePixelRatio: 1.0
logicalDpi: 96.0 x 96.0
physicalDpi: 135.5328125 x 135.46666666666664
Using calibre Qt style: True
[0.00] Starting up...
[0.00] Showing splash screen...
[1.05] splash screen shown
[1.05] Initializing db...
[1.06] db initialized
[1.06] Constructing main UI...
[2.44] main UI initialized...
[2.44] Hiding splash screen
Starting QuickView
DeDRM v7.2.1: Trying to decrypt 9781631230769.epub
DeDRM v7.2.1: Verifying zip archive integrity
DeDRM v7.2.1: “9781631230769.epub” is a secure Barnes & Noble ePub
DeDRM v7.2.1: Trying Encryption key mykey
Could not decrypt 4jt66jp6.epub because of an exception:
Traceback (most recent call last):
  File "calibre_plugins.dedrm.ignobleepub", line 266, in decryptBook
  File "calibre_plugins.dedrm.ignobleepub", line 197, in __init__
  File "calibre_plugins.dedrm.ignobleepub", line 168, in __init__
  File "Crypto\Cipher\AES.py", line 232, in new
  File "Crypto\Cipher\__init__.py", line 79, in _create_cipher
  File "Crypto\Cipher\_mode_cbc.py", line 274, in _create_cbc_cipher
  File "Crypto\Cipher\AES.py", line 93, in _create_base_cipher
ValueError: Incorrect AES key length (0 bytes)

DeDRM v7.2.1: Failed to decrypt with key mykey after 0.1 seconds
DeDRM v7.2.1: Looking for new NOOK Study Keys after 0.1 seconds
No nook Study log files have been found.
DeDRM v7.2.1: Ultimately failed to decrypt after 0.1 seconds. Read the FAQs at Harper's repository: https://github.com/apprenticeharper/DeDRM_tools/blob/master/FAQs.md
Running file type plugin DeDRM failed with traceback:
Traceback (most recent call last):
  File "calibre\customize\ui.py", line 175, in _run_filetype_plugins
  File "calibre_plugins.dedrm.__init__", line 644, in run
  File "calibre_plugins.dedrm.__init__", line 313, in ePubDecrypt
calibre_plugins.dedrm.DeDRMError: DeDRM v7.2.1: Ultimately failed to decrypt after 0.1 seconds. Read the FAQs at Harper's repository: https://github.com/apprenticeharper/DeDRM_tools/blob/master/FAQs.md
[18.28] splash screen hidden
[18.28] Started up in 18.28 seconds with 117 books
stdout+stderr from file dialog helper: [b'', b'']
piped data from file dialog helper: [b'\x84\xe98\xdc\x84Uk0n\xe6=\x9a f\xfc6R\x98\xba\x856\xcf\xa5P\x15t\xf4\x10\xc6\xcc\x8a\t', b'C:\\Users\\deliana\\AppData\\Local\\Packages\\BarnesNoble.Nook_ahnzqzva31enc\\LocalState\\9781631230769.epub']
Added Amaranthine Saga 05, FUMIKO v2 to db in: 0.1
Added 1 books in 0.6 seconds


I just tried to grab the key from the latest Android app on my emulator and it didn't seem to be stored in the same cchashdata.db file anymore, but tried an older version (v5.0.2.38 worked) and confirmed key had changed and DeDRM worked again after updating!

Thanks for letting me know this was possible, I had no idea the key could change.
